src/app/CommandSender.h

```diff
@@ -53,96 +53,158 @@ namespace app {
+         * - CHIP_ERROR*: All other cases.
+         *
+         * The CommandSender object MUST continue to exist after this call is completed. The application shall wait until it
+         * receives an OnDone call to destroy and free the object.
+         *
+         * @param[in] apCommandSender The command sender object that initiated the command transaction.
+         * @param[in] aError          A system error code that conveys the overall error code.
+         */
+        virtual void OnError(const CommandSender * apCommandSender, CHIP_ERROR aError) {}
+
+        /**
+         * OnDone will be called when CommandSender has finished all work and is safe to destroy and free the
+         * allocated CommandSender object.
+         *
+         * This function will:
+         *      - Always be called exactly *once* for a given CommandSender instance.
+         *      - Be called even in error circumstances.
+         *      - Only be called after a successful call to SendCommandRequest returns, if SendCommandRequest is used.
+         *      - Always be called before a successful return from SendGroupCommandRequest, if SendGroupCommandRequest is used.
+         *
+         * This function must be implemented to destroy the CommandSender object.
+         *
+         * @param[in] apCommandSender   The command sender object of the terminated invoke command transaction.
+         */
+        virtual void OnDone(CommandSender * apCommandSender) = 0;
+    };
+
+    // CommandSender::ExtendedCallback::OnResponse is public SDK API, so we cannot break
+    // source compatibility for it. To allow for additional values to be added at a future
+    // time without constantly changing the function's declaration parameter list, we are
+    // defining the struct AdditionalResponseData and adding that to the parameter list
+    // to allow for future extendability.
+    struct AdditionalResponseData
+    {
+        Optional<uint16_t> commandRef;
+    };
+
+    // CommandSender::ExtendedCallback::OnError is public SDK API, so we cannot break source
+    // compatibility for it. To allow for additional values to be added at a future time
+    // without constantly changing the function's declaration parameter list, we are
+    // defining the struct ErrorData and adding that to the parameter list
+    // to allow for future extendability.
+    struct ErrorData
+    {
+        /**
+         * The following errors will be delivered through chipError
+         *
+         * - CHIP_ERROR_TIMEOUT: A response was not received within the expected response timeout.
+         * - CHIP_ERROR_*TLV*: A malformed, non-compliant response was received from the server.
+         * - CHIP_ERROR encapsulating a StatusIB: If we got a non-path-specific
+         *   status response from the server.  In that case,
+         *   StatusIB::InitFromChipError can be used to extract the status.
+         * - CHIP_ERROR*: All other cases.
+         */
+        CHIP_ERROR chipError;
```

Comment:
```suggestion
        CHIP_ERROR error;
```
would be clearer, I think...